Sources/PointFree/Account/ChangeSeats.swift

```diff
@@ -0,0 +1,190 @@
+import HttpPipeline
+import HttpPipelineHtmlSupport
+import Optics
+import Prelude
+import Styleguide
+import Tuple
+
+// MARK: Middleware
+
+let confirmChangeSeatsResponse =
+  filterMap(require1 >>> pure, or: loginAndRedirect)
+    <<< fetchSeatsTaken
+    <<< requireStripeSubscription
+    <<< requireActiveSubscription
+    <<< requireTeamYearlySubscription
+    <| writeStatus(.ok)
+    >-> map(lower)
+    >>> respond(
+      view: confirmChangeSeatsView,
+      layoutData: { subscription, currentUser, seatsTaken in
+        SimplePageLayoutData(
+          currentUser: currentUser,
+          data: (subscription, currentUser, seatsTaken),
+          title: "Add or remove seats?"
+        )
+    }
+)
+
+let changeSeatsMiddleware =
+  filterMap(require1 >>> pure, or: loginAndRedirect)
+    <<< filterMap(
+      require2 >>> pure,
+      or: redirect(
+        to: .account(.subscription(.changeSeats(.show))),
+        headersMiddleware: flash(.error, "Couldn’t change the number of seats on your subscription.")
+      )
+    )
+    <<< fetchSeatsTaken
+    <<< requireStripeSubscription
+    <<< requireActiveSubscription
+    <<< requireTeamYearlySubscription
+    <<< requireValidSeating
+    <| changeSeats
+    >-> redirect(
+      to: .account(.index),
+      headersMiddleware: flash(.notice, "We’ve changed the number of seats on your subscription.")
+)
+
+// MARK: -
+
+private func changeSeats(_ conn: Conn<StatusLineOpen, Tuple4<Stripe.Subscription, Database.User, Int, Int>>)
+  -> IO<Conn<StatusLineOpen, Prelude.Unit>> {
+
+    let (subscription, _, seatsTaken, quantity) = lower(conn.data)
+
+    // TODO: send emails
+    return  AppEnvironment.current.stripe.updateSubscription(subscription, .teamYearly, quantity)
+      .map { sub -> Stripe.Subscription in
+        if sub.quantity > subscription.quantity {
+          parallel(AppEnvironment.current.stripe.invoiceCustomer(sub.customer).run)
+            .run({ _ in })
+        }
+
+        return sub
+      }
+      .run
+      .map(const(conn.map(const(unit))))
+}
+
+// MARK: - Transformers
+
+private func requireTeamYearlySubscription<A>(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, T3<Stripe.Subscription, Database.User, A>, Data>
+  )
+  -> Middleware<StatusLineOpen, ResponseEnded, T3<Stripe.Subscription, Database.User, A>, Data> {
+
+    return filter(
+      get1 >>> (^\.plan.id.unwrap == Stripe.Plan.Id.teamYearly.unwrap),
+      or: redirect(
+        to: .account(.index),
+        headersMiddleware: flash(.error, "You aren’t enrolled in a team subscription.")
+      )
+      )
+      <| middleware
+}
+
+private func fetchSeatsTaken<A>(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, T3<Database.User, Int, A>, Data>
+  )
+  -> Middleware<StatusLineOpen, ResponseEnded, T2<Database.User, A>, Data> {
+
+    return { conn -> IO<Conn<ResponseEnded, Data>> in
+      let user = conn.data.first
+
+      let invitesAndTeammates = sequence([
```

Comment:
More of this stuff could be parallelized and cleaned up in the future. If we add a proper query builder we can just `SELECT count(*)` here instead.